### ISSUE TITLE
Registering and Passing Variables Through A3C Graph

### ISSUE THREAD

**Issue by neale** [ppwwyyxx/tensorpack#192]
First, I really appreciate all the work that goes into maintaining this project that has so many models/examples that just run out of the box. 

I've been trying to augment the A3C gym network to pass recurrent states through the graph for each worker. I understand how it should work and I have an implementation that I think should work. But I'm hitting against how tensorpack deals with the graph and I can't get states to be managed correctly by the simulator. 

### Registering Layer

First I tried registering an LSTM/dynamic_rnn as a layer within tensorpack that looks like this
```python
import tensorflow as tf
from .common import layer_register

__all__ = ['Lstm']

@layer_register()
def Lstm(x, num_inputs):
    lstm_cell = tf.contrib.rnn.BasicLSTMCell(num_inputs, state_is_tuple=True)

    """
    ... 
    initialize lstm 
    """
    lstm_out, lstm_out_state = tf.nn.dynamic_rnn(lstm_cell, lstm_input, sequence_length,
                                                state_in,
                                                dtype=tf.float32)

    return (lstm_out, lstm_out_state)
    # other layers have a return name='output' but I couldn't find what this impacts. 
```

This layer doesn't actually register, and I can't access it externally. 

So I instead tried to roll the same layer with tensorflow in the model file, hoping that tensorflow/tensorpack ops all get handled in the same way at graph creation

### Tensorflow attempt

```python
def _get_NN_prediction(self, image):
   """
   ... rest of example A3C graph ...
   """
  with tf.variable_scope("lstm0") as scope:
            """
            ... init lstm 
            """
            #place LSTM within a retrievable scope to reuse hidden state
            lstm_cell = rnn.BasicLSTMCell(256, state_is_tuple=True)
            # same init
            lstm_out, lstm_out_state = tf.nn.dynamic_rnn(lstm_cell, lstm_input, sequence_length,
                                                             state_in,
                                                             dtype=tf.float32)
            scope.reuse_variables()
    # do policy FC layer
    # do value FC layer
    return policy, value
```

This will run (and never learn), but I don't know how to access/manipulate the LSTM within tensorpack. I think the LSTM weights are being reinitialized with every forward pass/never updating. 

I'm sure there is a succinct way to do this, in a way where the weights are accessible, since you need to freeze weights at time T_MAX. I tried branching to different LSTM layer with weights that are `trainable=False` with something like this 

Within `def_build_graph(...):`
```python
trainable = lambda: self._lstm_trainable(lstm_input) #original LSTM layer
frozen    = lambda: self._lstm_frozen(lstm_input) #copy of LSTM with non-trainable weights
pred = tf.cast(self.freeze, tf.bool) # is the network at T_MAX?
lstm_out  = tf.cond(pred, trainable, frozen) #choose one branch or the other
```
But that's hacky and has initialization issues. 

How can effectively register my layer in a way that I can manipulate it from the Simulator?

**Comment by ppwwyyxx** [comment]
1. If you added the layer to `tensorpack/models` as the first snippet, it should be accessible, just as any other layers. 
Maybe you've installed tensorpack through pip, so you are not running the actual code you're changing?
And you can also just register the layer in your own code and import it.

**Comment by ppwwyyxx** [comment]
2. This A3C implementation doesn't use T_MAX in a batch as the original paper, but a large batch with randomly shuffled experience, so simply adding a LSTM layer shouldn't work. 
It may need a continuous sequence in a batch, plus some way to manipulate states of each game simulator. Others have given a good way to do this: https://github.com/NVlabs/GA3C/issues/3#issuecomment-281247793

**Comment by ppwwyyxx** [comment]
A simple file like this placed in `models` is enough for `from tensorpack import *` to find `Lstm`:
```python
from .common import layer_register
__all__ = ['Lstm']
@layer_register()
def Lstm(x):
    pass
```

You can also manage the states by letting the predictor fetch the hidden state, and feeding the hidden state in training. This is supported under the framework as well. I mentioned this in that thread also, but I like the other solution.

**Comment by neale** [comment]
The probIem is that new layers don't seem to import past the top level. I did a fresh clone and pasted your example into tensorpack/tensorpack/models. 

At the  top level I get : 
```
>>> from tensorpack import models
>>> models.Conv2D
<function tensorpack.models.conv2d.Conv2D>
>>> models.Lstm
<function tensorpack.models.lstm.Lstm>
```
But as soon as I drop into `examples` or `A3C-Gym` I get an error
```
>>> from tensorpack import models
>>> models.Conv2D
<function tensorpack.models.conv2d.Conv2D>
>>> models.Lstm
'module' object has no attribute 'Lstm'
```

Thanks for your help, I just have one more question. I noticed that in the LSTM example you linked to, the solution utilizes lots of inputs like this
```python
self._inputvar_agent_indexs = InputDesc(tf.int32, [None], 'agent_index')
self._inputvar_R = InputDesc(tf.float32, [None, None], 'R')
self._inputvar_td = InputDesc(tf.float32, [None, None], 'td')
self._inputvar_is_over = InputDesc(tf.int32, [None], 'is_over')
self._inputvar_sequence_length = InputDesc(tf.int32, [None], 'sequence_length')
```
And these are obviously done by feeding dicts with `sess.run(states, feed_dict = {...})`

Does this play nicely with the trainer/predictor thats already going in `MySimulatorMaster`? Or does it overwrite everything, and the workers will need to be rewritten.

**Comment by ppwwyyxx** [comment]
1. Then it's very likely you `pip install` it, as I said above. Then any changes locally wouldn't appear unless you're in the top level directory.

2. What's input into the graph is determined here: https://github.com/ppwwyyxx/tensorpack/blob/master/examples/A3C-Gym/train-atari.py#L190.
If you want extra inputs, just add more InputDesc and change what's put into the queue.

**Comment by neale** [comment]
👍 got it, I had a virtualenv hanging around. I rebuilt the package and the layer registered fine. 

Ok I'll work through that, thanks a lot.
